scheds/rust/scx_bpfland/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,490 @@
+	 */
+	if (!builtin_idle)
+		return prev_cpu;
+
+	/*
+	 * Find another idle CPU and potentially migrate the task.
+	 */
+	cpu = scx_bpf_select_cpu_dfl(p, prev_cpu, wake_flags, &is_idle);
+	if (is_idle) {
+		dispatch_direct_cpu(p, cpu);
+		return cpu;
+	}
+
+	return prev_cpu;
+}
+
+void BPF_STRUCT_OPS(bpfland_enqueue, struct task_struct *p, u64 enq_flags)
+{
+	/*
+	 * Always dispatch per-CPU kthreads immediately.
+	 *
+	 * This allows to prioritize critical kernel threads that may
+	 * potentially slow down the entire system if they are blocked for too
+	 * long (i.e., ksoftirqd/N, rcuop/N, etc.).
+	 */
+	if (is_kthread(p) && p->nr_cpus_allowed == 1) {
+		dispatch_kthread(p, enq_flags);
+		return;
+	}
+
+	/*
+	 * Dispatch all the other tasks that were not dispatched directly in
+	 * select_cpu().
+	 */
+	dispatch_task(p, enq_flags);
+}
+
+void BPF_STRUCT_OPS(bpfland_dispatch, s32 cpu, struct task_struct *prev)
+{
+	/*
+	 * First consume directly dispatched tasks, so that they can
+	 * immediately use the CPU assigned in select_cpu().
+	 */
+	if (scx_bpf_consume(cpu_to_dsq(cpu)))
+		return;
+
+	/*
+	 * Then always consume interactive tasks before regular tasks.
+	 *
+	 * This is fine and we shouldn't have starvation, because interactive
+	 * tasks are classified by their amount of voluntary context switches,
+	 * so they should naturally release the CPU quickly and give a chance
+	 * to the regular tasks to run.
+	 */
+	if (scx_bpf_consume(PRIO_DSQ))
+		return;
+
+	/*
+	 * Lastly, consume regular tasks from the shared DSQ.
+	 */
+	scx_bpf_consume(SHARED_DSQ);
+}
+
+void BPF_STRUCT_OPS(bpfland_running, struct task_struct *p)
+{
+	/* Update global vruntime */
+	if (vtime_before(vtime_now, p->scx.dsq_vtime))
+		vtime_now = p->scx.dsq_vtime;
+
+	__sync_fetch_and_add(&nr_running, 1);
+}
+
+/*
+ * Update task statistics when the task is releasing the CPU (either
+ * voluntarily or because it expires its assigned time slice).
+ */
+void BPF_STRUCT_OPS(bpfland_stopping, struct task_struct *p, bool runnable)
+{
+	u64 slice = slice_ns - p->scx.slice;
+	u64 weight = p->scx.weight;
+	u64 now = bpf_ktime_get_ns();
+	struct task_ctx *tctx;
+
+	__sync_fetch_and_sub(&nr_running, 1);
+
+	tctx = lookup_task_ctx(p);
+	if (!tctx)
+		return;
+
+	/*
+	 * Classify interactive tasks based on the average amount of their
+	 * voluntary context switches and automatically boost their priority
+	 * (that will affect their vruntime).
+	 */
+	if (nvcsw_thresh && weight_boost) {
+		if (tctx->avg_nvcsw >= nvcsw_thresh)
+			tctx->is_interactive = true;
+		if (tctx->is_interactive)
+			weight *= weight_boost;
+	}
+
+	/*
+	 * Evaluate weighted time slice and assign the task a time slice within
+	 * the range [1, slice_ns]. Assigning a minimum slice of 1 ensures
+	 * strictly monotonic progress of the vruntime.
+	 */
+	slice = MAX(slice * 100 / weight, 1);
+	slice = MIN(slice, slice_ns);
+
+	/*
+	 * Update task's vruntime.
+	 */
+	p->scx.dsq_vtime += slice;
```

Comment:
It seems that `p->scx.dsq_vtime` is monotonically increasing. If it is true, does it mean a long-running task could be penalized over a freshly forked task? Am I correct?

Reply:
It should be the opposite, tasks are executed in order from the smallest vtime to the largest vtime. So we are actually penalizing the newly forked tasks. In fact one issue with this scheduler is that it's not very responsive at forking new tasks, like opening a new terminal session when the system is overloaded...

Reply:
Meh I think I got the vruntime logic backward... then we can definitely improve the scheduler, lol! @multics69 thank you so so much for your review! Will push a fix for this.